src/map/clif.cpp

```diff
@@ -20061,6 +20061,317 @@ void clif_achievement_reward_ack(int fd, unsigned char result, int achievement_i
+		count++;
+	}
+
+	// Return the amount of found materials
+	return count;
+}
+
+/**
+ * Adds the selected item into the refine UI and sends the possible materials
+ * to the client.
+ * 0aa2 <length>.W <index>.W <catalyst count>.B { <material>.W <chance>.B <price>.L }*
+ */
+void clif_refineui_info( struct map_session_data* sd, uint16 index ){
+	int fd = sd->fd;
+	struct item *item;
+	struct item_data *id;
+	uint16 length;
+	struct refine_materials materials[REFINEUI_MAT_CNT];
+	uint8 i, material_count;
+
+	// Get the item db reference
+	id = sd->inventory_data[index];
+
+	// No item data was found
+	if( id == NULL ){
+		return;
+	}
+
+	// Check if the item can be refined
+	if( id->flag.no_refine ){
+		return;
+	}
+
+	// Only weapons and armors can be refined in the refine UI
+	if( id->type != IT_WEAPON && id->type != IT_ARMOR ){
+		return;
+	}
+
+	// Check the inventory
+	item = &sd->inventory.u.items_inventory[index];
+
+	// No item was found at the given index
+	if( item == NULL ){
+		return;
+	}
+
+	// Check if the item is identified
+	if( !item->identify ){
+		return;
+	}
+
+	// Check if the item is broken
+	if( item->attribute ){
+		return;
+	}
+
+	// Calculate the possible materials
+	material_count = clif_refineui_materials( item, id, materials );
+
+	// No possibilities were found
+	if( material_count == 0 ){
+		return;
+	}
+
+	length = 7 + material_count * 7;
+
+	WFIFOHEAD(fd,length);
+	WFIFOW(fd,0) = 0x0AA2;
+	WFIFOW(fd,2) = length;
+	WFIFOW(fd,4) = index + 2;
+	WFIFOB(fd,6) = 0; //TODO: required amount of "Blacksmith Blessing"(id: 6635)
+
+	for( i = 0; i < material_count; i++ ){
+		WFIFOW(fd,7 + i * 7) = materials[i].cost.nameid;
+		WFIFOB(fd,7 + i * 7 + 2) = materials[i].chance;
+		WFIFOL(fd,7 + i * 7 + 3) = materials[i].cost.zeny;
+	}
+
+	WFIFOSET(fd,length);
+}
+
+/**
+ * Client request to add an item to the refine UI.
+ * 0aa1 <index>.W
+ */
+void clif_parse_refineui_add( int fd, struct map_session_data* sd ){
+#if PACKETVER >= 20161012
+	uint16 index = RFIFOW(fd, 2) - 2;
+
+	// Check if the refine UI is open
+	if( !sd->state.refineui_open ){
+		return;
+	}
+
+	// Check if the index is valid
+	if( index < 0 || index >= MAX_INVENTORY ){
+		return;
+	}
+
+	// Send out the requirements for the refine process
+	clif_refineui_info( sd, index );
+#endif
+}
+
+/**
+ * Client requests to try to refine an item.
+ * 0aa3 <index>.W <material>.W <catalyst>.B
+ */
+void clif_parse_refineui_refine( int fd, struct map_session_data* sd ){
+#if PACKETVER >= 20161012
+	uint16 index = RFIFOW( fd, 2 ) - 2;
+	uint16 material = RFIFOW( fd, 4 );
+	bool use_blacksmith_blessing = RFIFOB( fd, 6 ) != 0; // TODO: add logic
+	struct refine_materials materials[REFINEUI_MAT_CNT];
+	uint8 i, material_count;
+	uint16 j;
+	struct item *item;
+	struct item_data *id;
+
+	// Check if the refine UI is open
+	if( !sd->state.refineui_open ){
+		return;
+	}
+
+	// Get the item db reference
+	id = sd->inventory_data[index];
```

Comment:
add check for `index` maybe? to avoid WPE user crash the server.
also with `material` and `use_blacksmith_blessing` maybe.
if player has no material or BS blessing, force dc (mostly official disconnect player by sending invalid value)

Reply:
Thanks I missed that index check.
Material is item ID and is already checked below.
The logic for BS blessing is still missing, but will be added soon.